src/session_pool/session.js

```diff
@@ -0,0 +1,171 @@
+import moment from 'moment';
+
+import { checkParamOrThrow } from 'apify-client/build/utils';
+import EVENTS from './events';
+
+
+/**
+ *  Class aggregating data for `Session`.
+ *  Session internal state can be enriched with custom user data for example some authorization tokens.
+ */
+export default class Session {
+    constructor(options = {}) {
+        const {
+            id = cryptoRandomObjectId(),
+            cookies = [],
+            fingerprintSeed = cryptoRandomObjectId(),
+            maxAgeSecs = 3000,
+            userData = {},
+            maxErrorScore = 3,
+            errorScoreDecrement = 0.5,
+            createdAt = moment().toISOString(),
+            usageCount = 0,
+            errorScore = 0,
+            maxSessionUsageCount = 50,
+            sessionPool,
+        } = options;
+
+        let { expiresAt } = options;
+
+        // Validation
+        checkParamOrThrow(id, 'options.id', 'Maybe String');
+        checkParamOrThrow(cookies, 'options.cookies', 'Maybe Array');
+        checkParamOrThrow(fingerprintSeed, 'options.fingerprintSeed', 'Maybe String');
+        checkParamOrThrow(maxAgeSecs, 'options.maxAgeSecs', 'Maybe Number');
+        checkParamOrThrow(userData, 'options.userData', 'Maybe Object');
+        checkParamOrThrow(maxErrorScore, 'options.maxErrorScore', 'Maybe Number');
+        checkParamOrThrow(expiresAt, 'options.expiresAt', 'Maybe String');
+        checkParamOrThrow(createdAt, 'options.createdAt', 'Maybe String');
+        checkParamOrThrow(usageCount, 'options.usageCount', 'Maybe Number');
+        checkParamOrThrow(errorScore, 'options.errorScore', 'Maybe Number');
+        checkParamOrThrow(maxSessionUsageCount, 'options.maxSessionUsageCount', 'Maybe Number');
+        checkParamOrThrow(sessionPool, 'options.sessionPool', 'Object');
+
+        // sessionPool must be at least instance of EvenEmitter.
+        // That way we can allow custom implementation of SessionPool in the future (It should not be needed).
+        if (!(sessionPool instanceof EventEmitter)) {
+            throw new Error('Session: sessionPool must be instance of SessionPool');
+        }
+
+        if (!expiresAt) {
+            expiresAt = moment().add(maxAgeSecs, 'seconds').toISOString();
+        }
+
+        // Configurable
+        this.id = id;
+        this.cookies = cookies;
+        this.fingerprintSeed = fingerprintSeed;
+        this.maxAgeSecs = maxAgeSecs;
+        this.userData = userData;
+        this.maxErrorScore = maxErrorScore;
+        this.errorScoreDecrement = errorScoreDecrement; // TODO: Better Naming
+
+        // Internal
+        this.expiresAt = expiresAt;
+        this.createdAt = createdAt;
+        this.usageCount = usageCount;
+        this.errorScore = errorScore;
+        this.maxSessionUsageCount = maxSessionUsageCount;
+        this.sessionPool = sessionPool;
+    }
+
+    /**
+     * Decides whether the `Session` is blocked.
```

Comment:
Decides => Indicates 
`Session` => Session (no need to use code-like formatting)